tests/pytests/scenarios/failover/multimaster/test_failover_master.py

```diff
@@ -0,0 +1,306 @@
+import logging
+import time
+
+import pytest
+import salt.utils.platform
+from salt.utils.files import fopen
+from saltfactories.exceptions import FactoryTimeout
+
+pytestmark = [
+    pytest.mark.slow_test,
+    pytest.mark.skipif(
+        salt.utils.platform.is_freebsd(),
+        reason="Processes are not properly killed on FreeBSD",
+    ),
+]
+
+log = logging.getLogger(__name__)
+
+
+def _run_echo_for_all_possibilities(cli_list, minion_list):
+    """
+    Run test.echo from each cli to each minion.
+
+    Returns a list of minions that echoed back.
+    """
+    returned_minions = []
+
+    for cli in cli_list:
+        for minion in minion_list:
+            # Attempt to run test.echo from cli to the minion.
+            # If it the master has a key of an unconnected minion,  it will error out, so we handle it.
+            try:
+                ret = cli.run(
+                    "test.echo", "salt is cool!", minion_tgt=minion.id, _timeout=5
+                )
+                if ret and ret.json:
+                    assert ret.json == "salt is cool!"
+                    assert ret.exitcode == 0
+                    returned_minions.append(minion)
+            except FactoryTimeout as exc:
+                log.debug(
+                    "Failed to execute test.echo from %s to %s.",
+                    cli.get_display_name(),
+                    minion.id,
+                )
+
+    return returned_minions
+
+
+def test_pki(salt_mm_failover_minion_1):
+    """
+    Verify https://docs.saltproject.io/en/latest/topics/tutorials/multimaster_pki.html
+
+    We should keep this as the first test to minimize the size of the log file.
+    """
+    with fopen(salt_mm_failover_minion_1.config["log_file"]) as fp:
```

Comment:
Does the test need to be done against the log file?
We have a logging fixture, `caplog` which could achieve the same, unless, it needs to be tested against the log file.

Reply:
https://docs.pytest.org/en/stable/reference.html#caplog

And we have examples on our test suite using this fixture.

Reply:
Aha, what a useful fixture.  It has been swapped over to use that.